pkg/lease/v1alpha1/csp_lease.go

```diff
@@ -0,0 +1,254 @@
+// See the functions(below) for more details on update strategy
+func (sl *Lease) Update(podName string) (interface{}, error) {
+	newCspObject := sl.Object.(*apis.CStorPool)
+	if newCspObject.Annotations == nil {
+		sl.putKeyValue(podName, newCspObject)
+	} else if newCspObject.Annotations[sl.LeaseKey] == "" {
+		sl.putValue(podName, newCspObject)
+	} else {
+		sl.putUpdatedValue(podName, newCspObject)
+	}
+	csp, err := sl.Oecs.OpenebsV1alpha1().CStorPools().Update(newCspObject)
+	return csp, err
+}
+
+// Release method is implementation of  to release lease on a given csp.
+func (sl *Lease) Release() {
+	err := sl.patchCspLeaseAnnotation()
+	if err != nil {
+		newErr := fmt.Errorf("Lease could not be removed:%v", err)
+		runtime.HandleError(newErr)
+	}
+	glog.Info("Lease removed successfully on csp")
+}
+
+func (sl *Lease) getPodName() (string, error) {
+	podName := env.Get(PodName)
+	if strings.TrimSpace(podName) == "" {
+		return "", fmt.Errorf("Pod name not found in env variable")
+	}
+	podNameSpace := env.Get(NameSpace)
+	if strings.TrimSpace(podName) == "" {
+		return "", fmt.Errorf("Pod namespace not found in env variable")
+	}
+	return podNameSpace + "/" + podName, nil
+}
+
+// patchCspLeaseAnnotation will patch the lease key annotation on csp object to release the lease
+func (sl *Lease) patchCspLeaseAnnotation() error {
+	cspObject, ok := sl.Object.(*apis.CStorPool)
+	if !ok {
+		return fmt.Errorf("expected csp object for leasing but got %#v", cspObject)
+	}
+	leaseValueObj, err := parseLeaseValue(cspObject.Annotations[CspLeaseKey])
+	leaseValueObj.Holder = ""
+	newLeaseValue, err := json.Marshal(leaseValueObj)
+	if err != nil {
+		return err
+	}
+
+	cspPatch, err := patch.NewPatchPayload(PatchOperation, PatchPath, string(newLeaseValue))
+	if err != nil {
+		return fmt.Errorf("unable to form payload to patch csp:%s", err)
+	}
+	_, err = sl.Patch(cspObject.Name, "", types.JSONPatchType, cspPatch)
+	if err != nil {
+		return fmt.Errorf("unable to patch csp %s :%v", cspObject.Name, err)
+	}
+	return nil
+}
+
+// isLeaderLive checks whether the holder of lease is live or not
+// If the holder is not live or does not exists the function will return false.
+
+// If the holder of lease is not live or does not exists the lease can be acquired
+// by the other contestant(i.e. pool pod)
+func (sl *Lease) isLeaderALive(leaseValueObj LeaseContract) bool {
+	holderName := leaseValueObj.Holder
+	podDetails := strings.Split(holderName, "/")
+	// Check whether the holder is live or not.
+	pod, err := sl.Kubeclientset.CoreV1().Pods(podDetails[0]).Get(podDetails[1], meta_v1.GetOptions{})
+	if err != nil {
+		// If the pod does not exist, an error will be thrown and if it is a not found error
+		// meaning pod does not exist, we should return false.
+		if errors.IsNotFound(err) {
+			return false
+		}
+		return true
+	}
+	if pod == nil {
+		return false
+	}
+	podStatus := pod.Status.Phase
+	if string(podStatus) != string(v1.PodRunning) {
```

Comment:
@sonasingh46 -- This condition may not be sufficient. For example, the pod status can say "Terminating" - but the container in the pod may still be - running. Is there another option to fetch - what containers are running within a pod? For example -- podStatus != running and len(pod running containers) == 0, then return false.